### DISCUSSION TITLE
"Integrity Check Failed" -- Unclear what the issue is?

### DISCUSSION THREAD

**Discussion by JacobElder** [WinzarH/LikertMakeR#12] [Q&A] [answered]
Hi,

I love the package! It's very useful.

I just started trying to simulate some data using a correlation matrix, means, standard deviations and lower bounds. the code appears below. However, each time I run it, it simulates all 12 items and then fails saying that "Integrity Check Failed". Do you have any sense for what might be the cause of this issue and how to resolve?

```
n <- 500
structure(c(1, 0.330029948520833, 0.188608384038579, -0.227856832005643, 
0.0658065655590217, 0.361157559257308, -0.164031238154772, 0.334855411264458, 
0.321588501355856, 0.047050618877328, -5.58392988258599e-17, 
-0.110009982840277, 0.330029948520833, 1, 0.653971597517249, 
0.426130945042673, 0.122104194961678, 0.47866319213442, 0.366208848509599, 
0.658980836177279, 0.534712376510058, 0.37301978900801, 0.441872701025168, 
0.484536082474227, 0.188608384038579, 0.653971597517249, 1, 0.558684215397547, 
0.361593319019157, 0.422760021606695, 0.460425275902081, 0.576428719888396, 
0.438443850120786, 0.639529775414654, 0.515151515151515, 0.547922149271209, 
-0.227856832005643, 0.426130945042673, 0.558684215397547, 1, 
0.380859679057639, 0.505509334887457, 0.749711332475596, 0.425094947776568, 
0.479435297211452, 0.864096879880205, 0.787887996073464, 0.944172486074943, 
0.0658065655590217, 0.122104194961678, 0.361593319019157, 0.380859679057639, 
1, 0.35921425484608, 0.073020484067432, -0.0144805927201673, 
0.180789365841515, 0.373158800899798, 0.0951561365839888, 0.233108008563204, 
0.361157559257308, 0.47866319213442, 0.422760021606695, 0.505509334887457, 
0.35921425484608, 1, 0.240449258315369, 0.321672143858654, 0.316196436174188, 
0.524831265599411, 0.174077655955698, 0.449653301702031, -0.164031238154772, 
0.366208848509599, 0.460425275902081, 0.749711332475596, 0.073020484067432, 
0.240449258315369, 1, 0.477724234306496, 0.307903689187621, 0.829070229058699, 
0.739470897660918, 0.838211364366417, 0.334855411264458, 0.658980836177279, 
0.576428719888396, 0.425094947776568, -0.0144805927201673, 0.321672143858654, 
0.477724234306496, 1, 0.171857506542284, 0.610849840908602, 0.391971529524109, 
0.52449495124314, 0.321588501355856, 0.534712376510058, 0.438443850120786, 
0.479435297211452, 0.180789365841515, 0.316196436174188, 0.307903689187621, 
0.171857506542284, 1, 0.391761600772606, 0.67759504109576, 0.581209104902237, 
0.047050618877328, 0.37301978900801, 0.639529775414654, 0.864096879880205, 
0.373158800899798, 0.524831265599411, 0.829070229058699, 0.610849840908602, 
0.391761600772606, 1, 0.693957841407391, 0.896834811870322, -5.58392988258599e-17, 
0.441872701025168, 0.515151515151515, 0.787887996073464, 0.0951561365839888, 
0.174077655955698, 0.739470897660918, 0.391971529524109, 0.67759504109576, 
0.693957841407391, 1, 0.86607049400933, -0.110009982840277, 0.484536082474227, 
0.547922149271209, 0.944172486074943, 0.233108008563204, 0.449653301702031, 
0.838211364366417, 0.52449495124314, 0.581209104902237, 0.896834811870322, 
0.86607049400933, 1), dim = c(12L, 12L))
dfMeans <- c(6.16666666666667, 5.75, 5.75, 5.91666666666667, 5.58333333333333, 
5.75, 5.41666666666667, 5.75, 4.08333333333333, 5.58333333333333, 
5.5, 6.25)
dfSds <- c(0.834847109936722, 1.48477117918737, 0.866025403784439, 1.83195540504146, 
1.37895436890245, 1.05528970602217, 1.88092498199125, 1.13818036595899, 
1.97522534195852, 1.92865159365215, 1.73205080756888, 1.48477117918737
)
lowerbound <- rep(1, 12)
upperbound <- rep(7, 12)

simdf <- makeItems(
   n = n, 
   means = dfMeans, 
   sds = dfSds, 
   lowerbound = lowerbound, 
   upperbound = upperbound, 
   cormatrix = corMat
   )

```

**Comment by JacobElder** [comment]
It turns out if I round the values in the correlation matrix to the 2nd decimal, the simulation works. So it seems like the values in correlation matrix were too precise leading to an impossible solution, or something.

**Reply by WinzarH** [reply]
Requesting feedback on major changes to LikertMakeR package

Hello @JacobElder,



Thank you for your interest in — and contributions to — earlier versions of the LikertMakeR package.

I’ve made a substantial revision to the `makeCorrAlpha()` function, which generates a correlation matrix from a specified Cronbach’s alpha.

### Why the change?

In earlier versions, the function relied on a swap-based search procedure. While this worked well in most cases, it could be slow for larger numbers of items and occasionally produced matrices with more than one eigenvalue greater than 1. Because many users apply this function to reproduce published scales or generate teaching examples, I became concerned that the underlying structure did not consistently reflect a genuinely unidimensional scale.

### What’s new?

The revised function now constructs the correlation matrix directly from a single-factor model. Positive definiteness is guaranteed by construction, and the resulting matrices are inherently unidimensional.

As a result:

  - The function is dramatically faster (often more than 100× faster).

  - Positive-definiteness is ensured without post-hoc repair.

  - The generated structure better reflects how psychometric scales are typically designed.

However, some parameters have been removed and defaults have changed. This means the function is not fully backwards-compatible.

### I would greatly value your feedback before submitting this update to CRAN:

  - Does the new behaviour align with your expectations?

  - Do the parameter changes create issues in your existing workflows?

  - Can the documentation be improved?

  - Now that the generator enforces unidimensional structure, would you find value in a companion function for multi-factor correlation matrices?

The updated version is currently available on GitHub. Install with:

```
devtools::install_github("WinzarH/LikertMakeR")
```

Basic documentation is on the package website: <https://winzarh.github.io/LikertMakeR/reference/makeCorrAlpha.html>, and more detailed explanation is found in the website articles: <https://winzarh.github.io/LikertMakeR/articles/makeCorrAlpha_explainer.html>

I plan to leave the package there for a short period to gather feedback before CRAN submission. (What you can learn with a better grasp of matrix algebra!)

Thank you again, Jacob, for your engagement and support.

Hume

**Comment by WinzarH** [comment, marked as answer]
Hello Jacob,

Thanks for contacting me with the problem.

it's an intriguing issue.

The _"Integrity Check Failed"_ issue occurred in this case because the correlation matrix is not positive definite. That is, it is not feasible for creating correlated data.

The following is my output from your input parameters:


```
> print(corMat |> round(2))
       [,1] [,2] [,3]  [,4]  [,5] [,6]  [,7]  [,8] [,9] [,10] [,11] [,12]
 [1,]  1.00 0.33 0.19 -0.23  0.07 0.36 -0.16  0.33 0.32  0.05  0.00 -0.11
 [2,]  0.33 1.00 0.65  0.43  0.12 0.48  0.37  0.66 0.53  0.37  0.44  0.48
 [3,]  0.19 0.65 1.00  0.56  0.36 0.42  0.46  0.58 0.44  0.64  0.52  0.55
 [4,] -0.23 0.43 0.56  1.00  0.38 0.51  0.75  0.43 0.48  0.86  0.79  0.94
 [5,]  0.07 0.12 0.36  0.38  1.00 0.36  0.07 -0.01 0.18  0.37  0.10  0.23
 [6,]  0.36 0.48 0.42  0.51  0.36 1.00  0.24  0.32 0.32  0.52  0.17  0.45
 [7,] -0.16 0.37 0.46  0.75  0.07 0.24  1.00  0.48 0.31  0.83  0.74  0.84
 [8,]  0.33 0.66 0.58  0.43 -0.01 0.32  0.48  1.00 0.17  0.61  0.39  0.52
 [9,]  0.32 0.53 0.44  0.48  0.18 0.32  0.31  0.17 1.00  0.39  0.68  0.58
[10,]  0.05 0.37 0.64  0.86  0.37 0.52  0.83  0.61 0.39  1.00  0.69  0.90
[11,]  0.00 0.44 0.52  0.79  0.10 0.17  0.74  0.39 0.68  0.69  1.00  0.87
[12,] -0.11 0.48 0.55  0.94  0.23 0.45  0.84  0.52 0.58  0.90  0.87  1.00
> alpha(corMat) |> round(3)
[1] 0.897
> eigenvalues(corMat) |> round(3)
corMat  is NOT positive-definite

Eigenvalues:
 6.119888 1.813645 1.231597 0.9886005 0.6467144 0.5376285 0.2897258 0.2198587 0.1264369 0.01699969 0.008905852 -8.133767e-17 
 [1] 6.120 1.814 1.232 0.989 0.647 0.538 0.290 0.220 0.126 0.017 0.009 0.000
> simdf <- makeItems(
+   n = n,
+   means = dfMeans,
+   sds = dfSds,
+   lowerbound = lowerbound,
+   upperbound = upperbound,
+   cormatrix = corMat
+ )
Error in makeItems(n = n, means = dfMeans, sds = dfSds, lowerbound = lowerbound,  : 
  ERROR:
cormatrix is not Positive Definite.
         
Requested correlations are not possible 
> 

```

The error message "not Positive Definite" means that one or more of the eigenvalues of the correlation matrix is less than zero. Here the 12th eigenvalue is about -8.133767e-17, which I reckon is zero in any practical sense.  

So, thank you for alerting me to this case. I'm going to see if I can modify the code so that such values are rounded to zero (so long as they don't have other adverse effects).


I'm pleased that rounding values in the correaltion matrix helped your problem. But I was not able to replicate this solution. I still get the "not positive definite" error with rounding.

Can I ask two things:

What is the source of the correelation matrix and of the Means and Standard Deviations. They are unusually precise. I've never seen correlations reported to 15 decimal places!

Which version of _**LikertMakeR**_ are you using? The latest version **_(0.2.0)_** should have more detail in the error output - explaining that the correlation matrix is not positive definite, rather than "integrity check failed". Better still, download the development version **_(0.2.2)_**, to see a couple of new functions. It's available from GitHub:

```
     library(devtools)
     install_github("WinzarH/LikertMakeR")   

```
